README
NAME

    PlugAuth - Pluggable authentication and authorization server.

VERSION

    version 0.27

SYNOPSIS

    In the configuration file for the Clustericious app that will
    authenticate against PlugAuth:

     ---
     plug_auth:
       url: http://localhost:1234

    and authenticate and authorize in your Clustericious application's
    Routes.pm:

     authenticate;
     authorize;
     
     get '/resource' => sub {
       # resource that requires authentication
       # and authorization
     };

DESCRIPTION

    (For a quick start guide on how to setup a PlugAuth server, please see
    PlugAuth::Guide::Server)

    PlugAuth is a pluggable authentication and authorization server with a
    consistent RESTful API. This allows clients to authenticate and query
    authorization from a PlugAuth server without worrying or caring whether
    the actual authentication happens against flat files, PAM, LDAP or
    passed on to another PlugAuth server.

    The authentication API is HTTP Basic Authentication. The authorization
    API is based on users, groups, resources and hosts.

    The implementation for these can be swapped in and out depending on the
    plugins that you select in the configuration file. The default plugins
    for authentication (PlugAuth::Plugin::FlatAuth) and authorization
    (PlugAuth::Plugin::FlatAuthz) are implemented with ordinary flat files
    and advisory locks using flock.

    The are other plugins for ldap (PlugAuth::Plugin::LDAP), DBI
    (PlugAuth::Plugin::DBIAuth), or you can write your own
    (PlugAuth::Guide::Plugin).

    Here is a diagram that illustrates the most common use case for
    PlugAuth being used by a RESTful service:

      client
        |
        | HTTP
        |
     +-----------+          +------------+     +--------------+
     |   REST    |   HTTP   |            | --> | Auth Plugin  |  --> files
     |  service  |  ------> |  PlugAuth  |     +--------------+  --> ldap
     |           |          |            | --> | Authz Plugin |  --> ...
     +-----------+          +------------+     +--------------+

      1. Client (web browser or other) sends an HTTP request to the
      service.

      2. The service sends an HTTP basic authentication request to PlugAuth
      with the user's credentials

      3. PlugAuth performs authentication (see "AUTHENTICATION") and
      returns the appropriate HTTP status code.

      4. The REST service sends the HTTP status code to the client if
      authentication has failed.

      5. The REST service may optionally check the client's host, and if it
      is "trusted", authorization succeeds (see "AUTHORIZATION").

      6. If not, the REST service sends an authorization request to
      PlugAuth, asking whether the client has permission to perform an
      "action" on a "resource". Both the action and resource are arbitrary
      strings, though one reasonable default is sending the HTTP method as
      the action, and the URL path as the resource. (see "AUTHORIZATION"
      below).

      7. PlugAuth returns a response code to the REST service indicating
      whether or not authorization should succeed.

      8. The REST service returns the appropriate response to the client.

    If the REST service is written in Perl, see PlugAuth::Client.

    If the REST service uses Clustericious, see
    Clustericious::Plugin::PlugAuth.

    PlugAuth was originally written for scientific data processing clusters
    based on Clustericious in which all the services are RESTful servers
    distributed over a number of different physical hosts, though it may be
    applicable in other contexts.

 AUTHENTICATION

    Checking for authentication is done by sending a GET request to URLs of
    the form

     /auth

    With the username and password specified as HTTP Basic credentials. The
    actual mechanism used to verify authentication will depend on the
    authentication plugin being used. The default is
    PlugAuth::Plugin::FlatAuth.

 AUTHORIZATION

    Checking the authorization is done by sending GET requests to URLs of
    the form

     /authz/user/user/action/resource

    where user and action are strings (no slashes), and resource is a
    string which may have slashes. A response code of 200 indicates that
    access should be granted, 403 indicates that the resource is forbidden.
    A user is granted access to a resource if one of of the following
    conditions are met:

      * the user is specifically granted access to that resource, i.e. a
      line of the form

       /resource (action): username

      appears in the resources file (see "CONFIGURATION").

      * the user is a member of a group which is granted access to that
      resource.

      * the user or a group containing the user is granted access to a
      resource which is a prefix of the requested resource. i.e.

       / (action): username

      would grant access to "username" to perform "action" on any resource.

      * Additionally, given a user, an action, and a regular expression, it
      is possible to find all of the resources matching that regular
      expression for which the user has access. This can be done by sending
      a GET request to

       /authz/resources/user/action/regex

      * Host-based authorization is also possible -- sending a get request
      to

          /host/host/trusted

      where ".host" is a string representing a hostname, returns 200 if the
      host-based authorization should succeed, and 403 otherwise.

    For a complete list of the available routes and what they return see
    PlugAuth::Routes.

 CONFIGURATION

    Server configuration is done in ~/etc/PlugAuth.conf which is a
    Clustericious::Config style file. The configuration depends on which
    plugins you choose, consult your plugin's documentation. The default
    plugins are PlugAuth::Plugin::FlatAuth, PlugAuth::Plugin::FlatAuthz.

    Once the authentication and authorization has been configured, PlugAuth
    can be started (like any Mojolicious or Clustericious application)
    using the daemon command:

     % plugauth daemon

    This will use the built-in web server. To use another web server,
    additional configuration is required. For example, after adding this:

     start_mode: hypnotoad
     hypnotoad :
       listen : 'http://localhost:8099'
       env :
         %# Automatically generated configuration file
         HYPNOTOAD_CONFIG : /var/run/pluginauth/pluginauth.hypnotoad.conf

    This start command can be used to start a hypnotoad web server.

     % plugauth start

    See Clustericious::Config for more examples, including using with
    nginx, lighttpd, Plack or Apache.

EVENTS

    See PlugAuth::Routes

SEE ALSO

    Clustericious::Plugin::PlugAuth, PlugAuth::Client,
    PlugAuth::Guide::Client, PlugAuth::Guide::Plugin,
    PlugAuth::Guide::Server

AUTHOR

    Graham Ollis <gollis@sesda3.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2012 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.39      2018-03-11 14:30:57 -0400
  - Deprecated. See https://github.com/clustericious/Clustericious/issues/46

0.38      2017-07-26 04:31:05 -0400
  - Remove dependency on List::MoreUtils

0.37      2017-07-23 17:12:25 -0400
  - Require Clustericious 1.24 for recent Mojolicious compat

0.36      2017-07-22 21:59:57 -0400
  - Remove dependency on File::HomeDir

0.35      2016-11-28 14:31:26 -0500
  - Remove dependency on Log::Log4perl::CommandLine

0.34      2016-07-20 14:51:49 -0400
  - Mojo 7.x compatability

0.33      2015-12-07 19:54:31 -0500
  - Removed deprecated items in preperation for
    release of Clustericious 1.11.
  - Fixed test broken with current Clustericious dev
    (expected to be included with Clustericious 1.11)

0.32      2015-11-30 15:32:53 -0500
  - Remove deprecated use of JSON configuration files
    from test suite.
  - Remove dependency on Test::Differences
  - Migrate from Mojo::JSON to JSON::MaybeXS
  - Moved repository to clustericious org on github
    (updated meta appropriately)

0.31      2015-11-11 13:21:03 -0500
  - Silence warnings coming from FlatAuth
  - Fix bug where DisableGroup plugin would return 500
    error if the user was invalid

0.30      2015-03-13 03:53:43 -0400
  - Mojolicious 6.x compatability

0.29      2015-01-29 13:04:44 -0500
  - Recent Mojo compat test fix

0.28      2014-12-19 11:38:00 -0500
  - Merge PlugAuth::Plugin::Audit into this distro.
  - Merge PlugAuth::Client distro into this distro.

0.27      2014-12-09 10:24:00 America/New_York
  - use functional Mojo::JSON API
    the OO version was deprecated

0.26      2014-04-09 12:28:19 America/New_York
  - remove harness detection from test code

0.25      2014-02-24 10:29:37 America/New_York
  - alter logging, make authentication INFO
    instead of DEBUG

0.24      2014-01-22 17:57:53 America/New_York
  - Mojo 4.50 compat

0.23      2013-08-06 08:42:11 America/New_York
  - promote to production

0.22_01   2013-07-21 18:25:29 America/New_York
  - better diagnostics in new tests

0.22      2013-07-19 22:55:33 America/New_York
  - require Test::Clustericious::Cluster 0.08 for MSWin32 fixes

0.21      2013-07-19 16:19:45 America/New_York
  - promote to production

0.20_04   2013-07-19 11:14:19 America/New_York
  - allow create_user_cb to work even when implemented by a auth plugin NOT on the head

0.20_03   2013-07-19 09:20:15 America/New_York
  - added ability to assign groups to a user when user is created

0.20_02   2013-07-17 18:57:08 America/New_York
  - add disable_on_create option for DisableGroup Plugin

0.20_01   2013-07-17 15:49:31 America/New_York
  - add DisableGroup Plugin

0.20      2013-06-30 19:13:35 America/New_York
  - fix examples
  - use YAML::XS instead of YAML, since Clustericious requires YAML::XS (but not YAML)

0.19      2013-06-27 15:37:55 America/New_York
  - documentation
  - remove ladder and use before_dispatch hook instead for refresh

0.18      2013-06-14 10:50:23 America/New_York
  - take advantage of the new Test::Clustericious::Config

0.17      2013-05-30 15:05:49 America/New_York
  - Fixes for MSWin32

0.16      2013-05-17 22:39:37 America/New_York
  - Require Clustericious 0.9920

0.15      2013-05-17 20:14:41 America/New_York
  - Mojolicious 4.0 compatability

0.14      2013-05-10 11:57:36 EST5EDT
  - declare Class::Method::Modifiers as a prereq

0.12      2013-05-09 21:05:22 America/New_York
  - simplify the self auth plugin
  - don't use lib in main script
  - documentation updates

0.11      2013-04-23 16:53:33 America/New_York
  - use MakeMaker instead of ModuleBuild

0.10      2013-04-16 11:14:23 America/New_York
  - don't clobber user with route variables

0.09      2013-04-12 09:49:34 America/New_York
  - audit log for create,change,remove users,groups and resources
  - fix t/selfauth.t to not use config in ~/etc

0.08      2013-02-13 12:50:03 America/New_York
  - ensure the last line of the file has a new line in modificatons
  - ensure if the last line of the file doesn't have a new line adds still work

0.07      2012-12-27 18:05:54 America/New_York
  - return json if format is JSON.
  - return yml if format is YAML.
  - bug fix: /authz/user/username/action/ now returns 403 or 200 instead of always returning 404
  - feature: also allow /grant/user/username/action/
  - self auth against $c->authz instead of using Clustericious::Plugin::PlugAuth
  - Better tests for case insensitive usernames

0.06      2012-12-14 11:41:26 America/New_York
  - Avoid creating extra resources when using /authz/user/... routes
  - Created PlugAuth::Plugin::Test

0.05      2012-12-11 13:56:53 America/New_York
  - handle comments correctly for 'GET /grant'

0.04      2012-12-07 15:41:56 America/New_York
  - fixed /group/:group/#user route for users with @
  - added welcome role (PlugAuth::Role::Welcome)
  - wrote PlugAuth::Guide::Plugin (was a stub)

0.03      2012-12-03 10:43:56 America/New_York
  - wrote PlugAuth::Guide::Client (was a stub)
  - added grant route (GET /grant)

0.02      2012-11-29 15:13:48 America/New_York
  - Added specification tests for Plugins (Test::PlugAuth::Plugin::*)
  - use autodata to support JSON, YAML, whatever, instead of just JSON
  - use temp storage by default if no files are specified in the config
    warn on this sort of usage, as modifications will be lost.
  - Added revoke route (DELETE /grant/...)

0.01      2012-11-13 16:30:06 America/New_York
  - Added documentation in PlugAuth::Guide::*
  - Added roles for plugins (see PlugAuth::Role::*)
  - use Apache MD5 password when creating/changing passwords
  - accept Apache or Unix MD5 encrypted passwords
  - renamed PlugAuth from SimpleAuth
  - /users/:group returns 404 for groups that don't exist
  - POST /group/:group/:user to add a user to a group
  - DELETE /group/:group/:user to remove a user from a group


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Pluggable authentication and authorization server.",
  "AUTHOR" => "Graham Ollis <gollis\@sesda3.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "PlugAuth",
  "EXE_FILES" => [
    "bin/plugauth"
  ],
  "LICENSE" => "perl",
  "NAME" => "PlugAuth",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => 0,
    "Clone" => 0,
    "Clustericious" => "0.9936",
    "Clustericious::Config" => "0.25",
    "Clustericious::Log" => "0.14",
    "Crypt::PasswdMD5" => 0,
    "Exporter" => 0,
    "Fcntl" => 0,
    "File::HomeDir::Test" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "File::Touch" => 0,
    "File::stat" => 0,
    "FindBin" => 0,
    "IO::File" => 0,
    "List::MoreUtils" => 0,
    "Log::Log4perl" => 0,
    "Mojolicious" => "4.82",
    "Role::Tiny" => 0,
    "Test::Builder" => 0,
    "Text::Glob" => 0,
    "YAML::XS" => 0,
    "autodie" => 0
  },
  "TEST_REQUIRES" => {
    "File::HomeDir" => 0,
    "Test::Clustericious::Cluster" => "0.08",
    "Test::Clustericious::Config" => 0,
    "Test::Clustericious::Log" => 0,
    "Test::Differences" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.27",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => 0,
  "Clone" => 0,
  "Clustericious" => "0.9936",
  "Clustericious::Config" => "0.25",
  "Clustericious::Log" => "0.14",
  "Crypt::PasswdMD5" => 0,
  "Exporter" => 0,
  "Fcntl" => 0,
  "File::HomeDir" => 0,
  "File::HomeDir::Test" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "File::Touch" => 0,
  "File::stat" => 0,
  "FindBin" => 0,
  "IO::File" => 0,
  "List::MoreUtils" => 0,
  "Log::Log4perl" => 0,
  "Mojolicious" => "4.82",
  "Role::Tiny" => 0,
  "Test::Builder" => 0,
  "Test::Clustericious::Cluster" => "0.08",
  "Test::Clustericious::Config" => 0,
  "Test::Clustericious::Log" => 0,
  "Test::Differences" => 0,
  "Test::More" => 0,
  "Text::Glob" => 0,
  "YAML::XS" => 0,
  "autodie" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



